File: aeron-client/src/test/java/uk/co/real_logic/aeron/logbuffer/TermAppenderTest.java
/*
 * Copyright 2014 - 2015 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.aeron.logbuffer;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;
import uk.co.real_logic.aeron.protocol.DataHeaderFlyweight;
import uk.co.real_logic.agrona.concurrent.UnsafeBuffer;

import java.nio.ByteBuffer;

import static java.nio.ByteOrder.LITTLE_ENDIAN;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.*;
import static uk.co.real_logic.aeron.logbuffer.FrameDescriptor.*;
import static uk.co.real_logic.aeron.logbuffer.LogBufferDescriptor.TERM_META_DATA_LENGTH;
import static uk.co.real_logic.aeron.logbuffer.LogBufferDescriptor.TERM_TAIL_COUNTER_OFFSET;
import static uk.co.real_logic.aeron.logbuffer.TermAppender.TRIPPED;
import static uk.co.real_logic.aeron.protocol.DataHeaderFlyweight.HEADER_LENGTH;
import static uk.co.real_logic.agrona.BitUtil.*;

public class TermAppenderTest
{
    private static final int TERM_BUFFER_LENGTH = LogBufferDescriptor.TERM_MIN_LENGTH;
    private static final int META_DATA_BUFFER_LENGTH = TERM_META_DATA_LENGTH;
    private static final int MAX_FRAME_LENGTH = 1024;
    private static final int MAX_PAYLOAD_LENGTH = MAX_FRAME_LENGTH - HEADER_LENGTH;
    private static final UnsafeBuffer DEFAULT_HEADER = new UnsafeBuffer(ByteBuffer.allocateDirect(HEADER_LENGTH));
    private static final int TERM_ID = 7;

    private final UnsafeBuffer termBuffer = spy(new UnsafeBuffer(ByteBuffer.allocateDirect(TERM_BUFFER_LENGTH)));
    private final UnsafeBuffer metaDataBuffer = mock(UnsafeBuffer.class);

    private final HeaderWriter headerWriter =
        spy(new HeaderWriter(DataHeaderFlyweight.createDefaultHeader(0, 0, TERM_ID)));

    private TermAppender termAppender;

    @Before
    public void setUp()
    {
        when(termBuffer.capacity()).thenReturn(TERM_BUFFER_LENGTH);
        when(metaDataBuffer.capacity()).thenReturn(META_DATA_BUFFER_LENGTH);

        termAppender = new TermAppender(termBuffer, metaDataBuffer);
    }

    @Test
    public void shouldPackResult()
    {
        final int termId = 7;
        final int termOffset = -1;

        final long result = TermAppender.pack(termId, termOffset);

        assertThat(TermAppender.termId(result), is(termId));
        assertThat(TermAppender.termOffset(result), is(termOffset));
    }

    @Test
    public void shouldAppendFrameToEmptyLog()
    {
        final int headerLength = DEFAULT_HEADER.capacity();
        final UnsafeBuffer buffer = new UnsafeBuffer(new byte[128]);
        final int msgLength = 20;
        final int frameLength = msgLength + headerLength;
        final int alignedFrameLength = align(frameLength, FRAME_ALIGNMENT);
        final int tail = 0;

        when(metaDataBuffer.getAndAddLong(TERM_TAIL_COUNTER_OFFSET, alignedFrameLength))
            .thenReturn(TermAppender.pack(TERM_ID, tail));

        assertThat(termAppender.appendUnfragmentedMessage(headerWriter, buffer, 0, msgLength), is((long)alignedFrameLength));

        final InOrder inOrder = inOrder(termBuffer, metaDataBuffer, headerWriter);
        inOrder.verify(metaDataBuffer, times(1)).getAndAddLong(TERM_TAIL_COUNTER_OFFSET, alignedFrameLength);
        inOrder.verify(headerWriter, times(1)).write(termBuffer, tail, frameLength, TERM_ID);
        inOrder.verify(termBuffer, times(1)).putBytes(headerLength, buffer, 0, msgLength);
        inOrder.verify(termBuffer, times(1)).putIntOrdered(tail, frameLength);
    }

    @Test
    public void shouldAppendFrameTwiceToLog()
    {
        final int headerLength = DEFAULT_HEADER.capacity();
        final UnsafeBuffer buffer = new UnsafeBuffer(new byte[128]);
        final int msgLength = 20;
        final int frameLength = msgLength + headerLength;
        final int alignedFrameLength = align(frameLength, FRAME_ALIGNMENT);
        int tail = 0;

        when(metaDataBuffer.getAndAddLong(TERM_TAIL_COUNTER_OFFSET, alignedFrameLength))
            .thenReturn(TermAppender.pack(TERM_ID, tail))
            .thenReturn(TermAppender.pack(TERM_ID, alignedFrameLength));

        assertThat(termAppender.appendUnfragmentedMessage(
            headerWriter, buffer, 0, msgLength), is((long)alignedFrameLength));
        assertThat(termAppender.appendUnfragmentedMessage(
            headerWriter, buffer, 0, msgLength), is((long)alignedFrameLength * 2));

        final InOrder inOrder = inOrder(termBuffer, metaDataBuffer, headerWriter);
        inOrder.verify(metaDataBuffer, times(1)).getAndAddLong(TERM_TAIL_COUNTER_OFFSET, alignedFrameLength);
        inOrder.verify(headerWriter, times(1)).write(termBuffer, tail, frameLength, TERM_ID);
        inOrder.verify(termBuffer, times(1)).putBytes(headerLength, buffer, 0, msgLength);
        inOrder.verify(termBuffer, times(1)).putIntOrdered(tail, frameLength);

        tail = alignedFrameLength;
        inOrder.verify(metaDataBuffer, times(1)).getAndAddLong(TERM_TAIL_COUNTER_OFFSET, alignedFrameLength);
        inOrder.verify(headerWriter, times(1)).write(termBuffer, tail, frameLength, TERM_ID);
        inOrder.verify(termBuffer, times(1)).putBytes(tail + headerLength, buffer, 0, msgLength);
        inOrder.verify(termBuffer, times(1)).putIntOrdered(tail, frameLength);
    }

    @Test
    public void shouldPadLogAndTripWhenAppendingWithInsufficientRemainingCapacity()
    {
        final int msgLength = 120;
        final int headerLength = DEFAULT_HEADER.capacity();
        final int requiredFrameSize = align(headerLength + msgLength, FRAME_ALIGNMENT);
        final int tailValue = TERM_BUFFER_LENGTH - align(msgLength, FRAME_ALIGNMENT);
        final UnsafeBuffer buffer = new UnsafeBuffer(new byte[128]);
        final int frameLength = TERM_BUFFER_LENGTH - tailValue;

        when(metaDataBuffer.getAndAddLong(TERM_TAIL_COUNTER_OFFSET, requiredFrameSize))
            .thenReturn(TermAppender.pack(TERM_ID, tailValue));

        final long expectResult = TermAppender.pack(TERM_ID, TRIPPED);
        assertThat(termAppender.appendUnfragmentedMessage(headerWriter, buffer, 0, msgLength), is(expectResult));

        final InOrder inOrder = inOrder(termBuffer, metaDataBuffer, headerWriter);
        inOrder.verify(metaDataBuffer, times(1)).getAndAddLong(TERM_TAIL_COUNTER_OFFSET, requiredFrameSize);
        inOrder.verify(headerWriter, times(1)).write(termBuffer, tailValue, frameLength, TERM_ID);
        inOrder.verify(termBuffer, times(1)).putShort(typeOffset(tailValue), (short)PADDING_FRAME_TYPE, LITTLE_ENDIAN);
        inOrder.verify(termBuffer, times(1)).putIntOrdered(tailValue, frameLength);
    }

    @Test
    public void shouldPadLogAndTripWhenAppendingWithInsufficientRemainingCapacityIncludingHeader()
    {
        final int headerLength = DEFAULT_HEADER.capacity();
        final int msgLength = 120;
        final int requiredFrameSize = align(headerLength + msgLength, FRAME_ALIGNMENT);
        final int tailValue = TERM_BUFFER_LENGTH - (requiredFrameSize + (headerLength - FRAME_ALIGNMENT));
        final UnsafeBuffer buffer = new UnsafeBuffer(new byte[128]);
        final int frameLength = TERM_BUFFER_LENGTH - tailValue;

        when(metaDataBuffer.getAndAddLong(TERM_TAIL_COUNTER_OFFSET, requiredFrameSize))
            .thenReturn(TermAppender.pack(TERM_ID, tailValue));

        final long expectResult = TermAppender.pack(TERM_ID, TRIPPED);
        assertThat(termAppender.appendUnfragmentedMessage(headerWriter, buffer, 0, msgLength), is(expectResult));

        final InOrder inOrder = inOrder(termBuffer, metaDataBuffer, headerWriter);
        inOrder.verify(metaDataBuffer, times(1)).getAndAddLong(TERM_TAIL_COUNTER_OFFSET, requiredFrameSize);
        inOrder.verify(headerWriter, times(1)).write(termBuffer, tailValue, frameLength, TERM_ID);
        inOrder.verify(termBuffer, times(1)).putShort(typeOffset(tailValue), (short)PADDING_FRAME_TYPE, LITTLE_ENDIAN);
        inOrder.verify(termBuffer, times(1)).putIntOrdered(tailValue, frameLength);
    }

    @Test
    public void shouldFragmentMessageOverTwoFrames()
    {
        final int msgLength = MAX_PAYLOAD_LENGTH + 1;
        final int headerLength = DEFAULT_HEADER.capacity();
        final int frameLength = headerLength + 1;
        final int requiredCapacity = align(headerLength + 1, FRAME_ALIGNMENT) + MAX_FRAME_LENGTH;
        final UnsafeBuffer buffer = new UnsafeBuffer(new byte[msgLength]);
        int tail  = 0;

        when(metaDataBuffer.getAndAddLong(TERM_TAIL_COUNTER_OFFSET, requiredCapacity))
            .thenReturn(TermAppender.pack(TERM_ID, tail));

        assertThat(termAppender.appendFragmentedMessage(
            headerWriter, buffer, 0, msgLength, MAX_PAYLOAD_LENGTH), is((long)requiredCapacity));

        final InOrder inOrder = inOrder(termBuffer, metaDataBuffer, headerWriter);
        inOrder.verify(metaDataBuffer, times(1)).getAndAddLong(TERM_TAIL_COUNTER_OFFSET, requiredCapacity);

        inOrder.verify(headerWriter, times(1)).write(termBuffer, tail, MAX_FRAME_LENGTH, TERM_ID);
        inOrder.verify(termBuffer, times(1)).putBytes(tail + headerLength, buffer, 0, MAX_PAYLOAD_LENGTH);
        inOrder.verify(termBuffer, times(1)).putByte(flagsOffset(tail), BEGIN_FRAG_FLAG);
        inOrder.verify(termBuffer, times(1)).putIntOrdered(tail, MAX_FRAME_LENGTH);

        tail = MAX_FRAME_LENGTH;
        inOrder.verify(headerWriter, times(1)).write(termBuffer, tail, frameLength, TERM_ID);
        inOrder.verify(termBuffer, times(1)).putBytes(tail + headerLength, buffer, MAX_PAYLOAD_LENGTH, 1);
        inOrder.verify(termBuffer, times(1)).putByte(flagsOffset(tail), END_FRAG_FLAG);
        inOrder.verify(termBuffer, times(1)).putIntOrdered(tail, frameLength);
    }

    @Test
    public void shouldClaimRegionForZeroCopyEncoding()
    {
        final int headerLength = DEFAULT_HEADER.capacity();
        final int msgLength = 20;
        final int frameLength = msgLength + headerLength;
        final int alignedFrameLength = align(frameLength, FRAME_ALIGNMENT);
        final int tail = 0;
        final BufferClaim bufferClaim = new BufferClaim();

        when(metaDataBuffer.getAndAddLong(TERM_TAIL_COUNTER_OFFSET, alignedFrameLength))
            .thenReturn(TermAppender.pack(TERM_ID, tail));

        assertThat(termAppender.claim(headerWriter, msgLength, bufferClaim), is((long)alignedFrameLength));

        assertThat(bufferClaim.offset(), is(tail + headerLength));
        assertThat(bufferClaim.length(), is(msgLength));

        // Map flyweight or encode to buffer directly then call commit() when done
        bufferClaim.commit();

        final InOrder inOrder = inOrder(metaDataBuffer, headerWriter);
        inOrder.verify(metaDataBuffer, times(1)).getAndAddLong(TERM_TAIL_COUNTER_OFFSET, alignedFrameLength);
        inOrder.verify(headerWriter, times(1)).write(termBuffer, tail, frameLength, TERM_ID);
    }
}

File: aeron-client/src/test/cpp/concurrent/TermAppenderTest.cpp
/*
 * Copyright 2014-2018 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <array>

#include <gtest/gtest.h>

#include <thread>
#include "MockAtomicBuffer.h"
#include <concurrent/logbuffer/TermAppender.h>
#include <concurrent/logbuffer/ExclusiveTermAppender.h>

using namespace aeron::concurrent::logbuffer;
using namespace aeron::concurrent::mock;
using namespace aeron::concurrent;
using namespace aeron;

#define TERM_BUFFER_CAPACITY (LogBufferDescriptor::TERM_MIN_LENGTH)
#define META_DATA_BUFFER_CAPACITY (LogBufferDescriptor::LOG_META_DATA_LENGTH)
#define MAX_FRAME_LENGTH (1024)

#define MAX_PAYLOAD_LENGTH ((MAX_FRAME_LENGTH - DataFrameHeader::LENGTH))
#define SRC_BUFFER_CAPACITY (2 * 1024)
#define TERM_ID (101)
#define RESERVED_VALUE (777L)
#define PARTITION_INDEX (1)
#define TERM_TAIL_OFFSET (LogBufferDescriptor::TERM_TAIL_COUNTER_OFFSET + (PARTITION_INDEX * sizeof(std::int64_t)))

typedef std::array<std::uint8_t, TERM_BUFFER_CAPACITY> term_buffer_t;
typedef std::array<std::uint8_t, META_DATA_BUFFER_CAPACITY> meta_data_buffer_t;
typedef std::array<std::uint8_t, DataFrameHeader::LENGTH> hdr_t;
typedef std::array<std::uint8_t, SRC_BUFFER_CAPACITY> src_buffer_t;

static std::int64_t packRawTail(std::int32_t termId, std::int32_t termOffset)
{
    return static_cast<std::int64_t>(termId) << 32 | termOffset;
}

static std::int64_t reservedValueSupplier(AtomicBuffer&, util::index_t, util::index_t)
{
    return RESERVED_VALUE;
}

class TermAppenderTest : public testing::Test
{
public:
    TermAppenderTest() :
        m_termBuffer(m_logBuffer.data(), m_logBuffer.size()),
        m_metaDataBuffer(m_stateBuffer.data(), m_stateBuffer.size()),
        m_hdr(m_hdrBuffer, 0),
        m_src(m_srcBuffer, 0),
        m_headerWriter(m_hdr),
        m_termAppender(m_termBuffer, m_metaDataBuffer, PARTITION_INDEX)
    {
        m_logBuffer.fill(0);
        m_stateBuffer.fill(0);
    }

    virtual void SetUp()
    {
        m_logBuffer.fill(0);
        m_stateBuffer.fill(0);
        m_hdrBuffer.fill(0);
        m_srcBuffer.fill(0);
    }

protected:
    AERON_DECL_ALIGNED(term_buffer_t m_logBuffer, 16);
    AERON_DECL_ALIGNED(meta_data_buffer_t m_stateBuffer, 16);
    AERON_DECL_ALIGNED(hdr_t m_hdrBuffer, 16);
    AERON_DECL_ALIGNED(src_buffer_t m_srcBuffer, 16);
    MockAtomicBuffer m_termBuffer;
    MockAtomicBuffer m_metaDataBuffer;
    AtomicBuffer m_hdr;
    AtomicBuffer m_src;
    HeaderWriter m_headerWriter;
    TermAppender m_termAppender;
};

TEST_F(TermAppenderTest, shouldReportCapacity)
{
    EXPECT_EQ(m_termAppender.termBuffer().capacity(), TERM_BUFFER_CAPACITY);
}

TEST_F(TermAppenderTest, shouldAppendFrameToEmptyLog)
{
    const std::int32_t msgLength = 20;
    const std::int32_t frameLength = DataFrameHeader::LENGTH + msgLength;
    const std::int64_t alignedFrameLength = util::BitUtil::align(frameLength, FrameDescriptor::FRAME_ALIGNMENT);
    util::index_t tail = 0;
    testing::Sequence sequence;

    EXPECT_CALL(m_metaDataBuffer, getAndAddInt64(TERM_TAIL_OFFSET, alignedFrameLength))
        .Times(1)
        .InSequence(sequence)
        .WillOnce(testing::Return(packRawTail(TERM_ID, tail)));

    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), -frameLength))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putBytes(DataFrameHeader::LENGTH, testing::Ref(m_src), 0, msgLength))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putInt64(tail + DataFrameHeader::RESERVED_VALUE_FIELD_OFFSET, RESERVED_VALUE))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), frameLength))
        .Times(1)
        .InSequence(sequence);

    const std::int32_t resultingOffset =
        m_termAppender.appendUnfragmentedMessage(m_headerWriter, m_src, 0, msgLength, reservedValueSupplier, TERM_ID);
    EXPECT_EQ(resultingOffset, alignedFrameLength);
}

TEST_F(TermAppenderTest, shouldAppendFrameTwiceToLog)
{
    const util::index_t msgLength = 20;
    const util::index_t frameLength = DataFrameHeader::LENGTH + msgLength;
    const util::index_t alignedFrameLength = util::BitUtil::align(frameLength, FrameDescriptor::FRAME_ALIGNMENT);
    util::index_t tail = 0;
    testing::Sequence sequence1;
    testing::Sequence sequence2;

    EXPECT_CALL(m_metaDataBuffer, getAndAddInt64(TERM_TAIL_OFFSET, alignedFrameLength))
        .Times(2)
        .WillOnce(testing::Return(packRawTail(TERM_ID, tail)))
        .WillOnce(testing::Return(packRawTail(TERM_ID, alignedFrameLength)));

    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), -frameLength))
        .Times(1)
        .InSequence(sequence1);
    EXPECT_CALL(m_termBuffer, putBytes(tail + DataFrameHeader::LENGTH, testing::Ref(m_src), 0, msgLength))
        .Times(1)
        .InSequence(sequence1);
    EXPECT_CALL(m_termBuffer, putInt64(tail + DataFrameHeader::RESERVED_VALUE_FIELD_OFFSET, RESERVED_VALUE))
        .Times(1)
        .InSequence(sequence1);
    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), frameLength))
        .Times(1)
        .InSequence(sequence1);

    std::int32_t resultingOffset =
        m_termAppender.appendUnfragmentedMessage(m_headerWriter, m_src, 0, msgLength, reservedValueSupplier, TERM_ID);
    EXPECT_EQ(resultingOffset, alignedFrameLength);

    tail = alignedFrameLength;

    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), -frameLength))
        .Times(1)
        .InSequence(sequence2);
    EXPECT_CALL(m_termBuffer, putBytes(tail + DataFrameHeader::LENGTH, testing::Ref(m_src), 0, msgLength))
        .Times(1)
        .InSequence(sequence2);
    EXPECT_CALL(m_termBuffer, putInt64(tail + DataFrameHeader::RESERVED_VALUE_FIELD_OFFSET, RESERVED_VALUE))
        .Times(1)
        .InSequence(sequence2);
    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), frameLength))
        .Times(1)
        .InSequence(sequence2);

    resultingOffset =
        m_termAppender.appendUnfragmentedMessage(m_headerWriter, m_src, 0, msgLength, reservedValueSupplier, TERM_ID);
    EXPECT_EQ(resultingOffset, alignedFrameLength * 2);
}

TEST_F(TermAppenderTest, shouldPadLogWhenAppendingWithInsufficientRemainingCapacity)
{
    const util::index_t msgLength = 120;
    const util::index_t requiredFrameSize = util::BitUtil::align(msgLength + DataFrameHeader::LENGTH, FrameDescriptor::FRAME_ALIGNMENT);
    const util::index_t tailValue = TERM_BUFFER_CAPACITY - util::BitUtil::align(msgLength, FrameDescriptor::FRAME_ALIGNMENT);
    const util::index_t frameLength = TERM_BUFFER_CAPACITY - tailValue;
    testing::Sequence sequence;

    EXPECT_CALL(m_metaDataBuffer, getAndAddInt64(TERM_TAIL_OFFSET, requiredFrameSize))
        .Times(1)
        .WillOnce(testing::Return(packRawTail(TERM_ID, tailValue)));

    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tailValue), -frameLength))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putUInt16(FrameDescriptor::typeOffset(tailValue), DataFrameHeader::HDR_TYPE_PAD))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tailValue), frameLength))
        .Times(1)
        .InSequence(sequence);

    const std::int32_t resultingOffset =
        m_termAppender.appendUnfragmentedMessage(m_headerWriter, m_src, 0, msgLength, reservedValueSupplier, TERM_ID);
    EXPECT_EQ(resultingOffset, TERM_APPENDER_FAILED);
}

TEST_F(TermAppenderTest, shouldFragmentMessageOverTwoFrames)
{
    const util::index_t msgLength = MAX_PAYLOAD_LENGTH + 1;
    const util::index_t frameLength = DataFrameHeader::LENGTH + 1;
    const std::int64_t requiredCapacity = util::BitUtil::align(frameLength, FrameDescriptor::FRAME_ALIGNMENT) + MAX_FRAME_LENGTH;
    util::index_t tail = 0;
    testing::Sequence sequence;

    EXPECT_CALL(m_metaDataBuffer, getAndAddInt64(TERM_TAIL_OFFSET, requiredCapacity))
        .Times(1)
        .InSequence(sequence)
        .WillOnce(testing::Return(packRawTail(TERM_ID, tail)));

    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), -MAX_FRAME_LENGTH))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putBytes(tail + DataFrameHeader::LENGTH, testing::Ref(m_src), 0, MAX_PAYLOAD_LENGTH))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putUInt8(FrameDescriptor::flagsOffset(tail), FrameDescriptor::BEGIN_FRAG))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putInt64(tail + DataFrameHeader::RESERVED_VALUE_FIELD_OFFSET, RESERVED_VALUE))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), MAX_FRAME_LENGTH))
        .Times(1)
        .InSequence(sequence);

    tail = MAX_FRAME_LENGTH;

    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), -frameLength))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(
        m_termBuffer, putBytes(tail + DataFrameHeader::LENGTH, testing::Ref(m_src), MAX_PAYLOAD_LENGTH, 1))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putUInt8(FrameDescriptor::flagsOffset(tail), FrameDescriptor::END_FRAG))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putInt64(tail + DataFrameHeader::RESERVED_VALUE_FIELD_OFFSET, RESERVED_VALUE))
        .Times(1)
        .InSequence(sequence);
    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), frameLength))
        .Times(1)
        .InSequence(sequence);

    const std::int32_t resultingOffset =
        m_termAppender.appendFragmentedMessage(
            m_headerWriter, m_src, 0, msgLength, MAX_PAYLOAD_LENGTH, reservedValueSupplier, TERM_ID);
    EXPECT_EQ(resultingOffset, requiredCapacity);
}

TEST_F(TermAppenderTest, shouldClaimRegionForZeroCopyEncoding)
{
    const util::index_t msgLength = 20;
    const util::index_t frameLength = DataFrameHeader::LENGTH + msgLength;
    const std::int64_t alignedFrameLength = util::BitUtil::align(frameLength, FrameDescriptor::FRAME_ALIGNMENT);
    util::index_t tail = 0;
    BufferClaim bufferClaim;
    testing::Sequence sequence;

    EXPECT_CALL(m_metaDataBuffer, getAndAddInt64(TERM_TAIL_OFFSET, alignedFrameLength))
        .Times(1)
        .InSequence(sequence)
        .WillOnce(testing::Return(packRawTail(TERM_ID, tail)));

    EXPECT_CALL(m_termBuffer, putInt32Ordered(FrameDescriptor::lengthOffset(tail), -frameLength))
        .Times(1)
        .InSequence(sequence);

    const std::int32_t resultingOffset =
        m_termAppender.claim(m_headerWriter, msgLength, bufferClaim, TERM_ID);
    EXPECT_EQ(resultingOffset, alignedFrameLength);

    EXPECT_EQ(bufferClaim.offset(), (tail + DataFrameHeader::LENGTH));
    EXPECT_EQ(bufferClaim.length(), msgLength);

    bufferClaim.commit();
}
